server/lib/chat/chat-models.ts

```diff
@@ -578,3 +580,120 @@ export async function findChannelByName(
     done()
   }
 }
+
+/**
+ * Returns a list of non-private chat channels, optionally filtered by a `searchStr`. The list
+ * doesn't include the user's joined channel by default, but can be configured to include them as
+ * well (including the joined private channels).
+ */
+export async function searchChannelsAsUser(
+  {
+    userId,
+    limit,
+    pageNumber,
+    searchStr,
+    joined,
+  }: {
+    /** User requesting the list of channels. */
+    userId: SbUserId
+    /** Amount of channels to return. */
+    limit: number
+    /** Offset from which to start listing the channels. */
+    pageNumber: number
+    /** String to filter the channels by. */
+    searchStr?: string
+    /** Flag which includes the user's joined channels if true. */
+    joined?: boolean
+  },
+  withClient?: DbClient,
+): Promise<ChannelInfo[]> {
+  const { client, done } = await db(withClient)
+  try {
+    const query = joined
+      ? sql`
+        WITH joined_channels AS (
+          SELECT DISTINCT(channel_id)
+          FROM channel_users
+          WHERE channel_id IN (SELECT channel_id FROM channel_users WHERE user_id = ${userId})
+        )
+        SELECT c.*
+        FROM channels c
+        INNER JOIN joined_channels ON c.id = joined_channels.channel_id
+
+        UNION
+
+        SELECT *
+        FROM channels
+      `
+      : sql`
+        WITH unjoined_channels AS (
+          SELECT DISTINCT(channel_id)
+          FROM channel_users
+          WHERE channel_id NOT IN (SELECT channel_id FROM channel_users WHERE user_id = ${userId})
+        )
+        SELECT c.*
+        FROM channels c
+        INNER JOIN unjoined_channels ON c.id = unjoined_channels.channel_id
+      `
+
+    query.append(sql`WHERE private = false`)
+    if (searchStr) {
```

Comment:
This filter doesn't apply to the joined channels, which just seems broken? Instead of trying to build up this complex series of appends to generate the query you want without duplicating code, I think you would get a lot of value out of just writing the 2 queries you want straightforwardly, with as few conditional statements as possible. I recognize that you will repeat some code to do that, but I promise it is better than trying to figure out all the possible queries that can be generated by this current code.